kafka-client/src/main/java/dev/responsive/kafka/store/CommitBuffer.java

```diff
@@ -405,17 +385,17 @@ private AtomicWriteResult flush(final long offset, final UUID txnid, final int b
       return writeResult;
     }
 
-    writeResult = drain(writers.values(), AtomicWriter::finalizeTxn);
+    // this offset is only used for recovery, so it can (and should) be done only
+    // when all the flushes above have completed
+    writeResult = drain(writers.values(), writer -> writer.setOffset(offset));
```

Comment:
We can just track the offset in the base partition now that it's just a recovery optimization. Might help reduce some write load. Not something we need to do in this PR though.

Reply:
will do in a follow-up - I actually think this could be really helpful to reduce the write load (it'll cut it in half)